subprojects/docs/src/docs/release/notes.md

```diff
@@ -78,6 +78,48 @@ The `buildSrc` build of the root project continues to use its own build cache co
 However, the `buildSrc` build of any included build will inherit the build cache configuration from the root build.
 For more on configuring the build cache for the root `buildSrc` build, please see [the Userguide section on this topic](userguide/build_cache.html#buildCacheBuildSrc).
 
+### Richer task inputs
+
+The detection of task input and output annotations for `@Nested` properties now works on the runtime type of the nested bean.
+Earlier Gradle versions used the declared type for detecting annotations.
+Moreover, `Iterable`s can now be annotated with `@Nested` and each element of the iterable will be added as a separate nested input.
+
+This allows for richer modelling inputs and solves some problems on how to declare inputs which are only determined by subclasses.
+
+### Rich Java compiler arguments
+
+As a first use-case of richer task inputs, the `CompileJava` task now supports richer compiler arguments.
+It declares a property [`CompileJava.options.compilerArgumentProviders`](dsl/org.gradle.api.tasks.compile.CompileOptions.html#org.gradle.api.tasks.compile.CompileOptions:compilerArgumentProviders), which allows to add instances implementing [`CompilerArgumentProvider`](javadoc/org/gradle/api/tasks/compile/CompilerArgumentProvider.html) and then adds the necessary command line arguments.
```

Comment:
This is difficult stuff to explain. We could lead with an example, and restrict the scope to annotation processors:

> When you have to expose a file location to your annotation processor, it is essential for Gradle to learn about this additional input (or output). Without tracking the location and contents of the given file (or directory), features like incremental build and task output caching cannot function correctly. Before Gradle 4.5 you had to let Gradle know manually by calling `compileJava.inputs.file(...)` or similar.
> 
> Gradle 4.5 introduces a better way to handle this situation by modeling the annotation processor as a `CompilerArgumentProvider`. This approach allows the declaration of complex inputs and outputs, just like how you would declare `@InputFile` and `@OutputDirectory` properties on the task type.

I'd add the example code here, and then finish by saying that this method can be used to declare any kind of command-line argument for the compiler, not just annotation processor arguments.